api/src/main/java/org/xnio/http/HttpUpgrade.java

```diff
@@ -525,4 +526,9 @@ public void handleCancelled(Object attachment) {
             }
         }
     }
+    public static class ConnectionClosedEarlyException extends EOFException {
```

Comment:
Exception classes should not be nested classes, and they also need to have serialization information, JavaDoc, and appropriate constructors.  I'd make this a top level class and make it extend `UpgradeFailedException` instead.

Reply:
Thanks, updated now. If it looks good I'll update the 3.7 and ejb-client PRs on Monday